mesh/spacing.hpp

```diff
@@ -0,0 +1,689 @@
+// Copyright (c) 2010-2023, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_SPACING
+#define MFEM_SPACING
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+typedef enum {UNIFORM_SPACING, LINEAR, GEOMETRIC, BELL,
+              GAUSSIAN, LOGARITHMIC, PIECEWISE
+             } SPACING_TYPE;
+
+/// Class for spacing functions that define meshes in a dimension, using a
+/// formula or method implemented in a derived class.
+class SpacingFunction
+{
+public:
+   /** @brief Base class constructor.
+   @param[in] n_  Size or number of intervals, which defines elements.
+   @param[in] r   Whether to reverse the spacings, false by default.
+   @param[in] s   Whether to scale parameters by the refinement or coarsening
+                  factor, in the function @a SpacingFunction::ScaleParameters.
+   */
+   SpacingFunction(int n_, bool r=false, bool s=false)
+   {
+      n = n_;
+      reverse = r;
+      scale = s;
+   }
+
+   /// Returns the size, or number of intervals (elements).
+   inline int Size() const { return n; }
+
+   /// Sets the size, or number of intervals (elements).
+   virtual void SetSize(int size) = 0;
+
+   /// Sets the property that determines whether the spacing is reversed.
+   void SetReverse(bool r) { reverse = r; }
+
+   /// Returns the width of interval @a p (between 0 and @a Size() - 1).
+   virtual double Eval(int p) = 0;
+
+   /// Returns the width of all intervals, resizing @a s to @a Size().
+   void EvalAll(Vector & s)
+   {
+      s.SetSize(n);
+      for (int i=0; i<n; ++i)
+      {
+         s[i] = Eval(i);
+      }
+   }
+
+   /** @brief Scales parameters by the factor @a a associated with @a Size().
+
+       Note that parameters may be scaled inversely during coarsening and
+       refining, so the scaling should be linear in the sense that scaling by a
+       number followed by scaling by its inverse has no effect on parameters. */
+   virtual void ScaleParameters(double a) { }
```

Comment:
Some types of spacing functions have no parameters to scale, by definition, so this function should be a no-op for them. I don't think there should be a warning in any case. There are 2 options: (i) have the no-op in the base class as done already; (ii) make the method pure virtual and put the no-op implementation in the derived classes. I chose (i) to avoid reimplementing the no-op in multiple derived classes, and I don't see an advantage to (ii).